test/e2e/storage/csi_volumes.go

```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
+	},
+		"test/e2e/testing-manifests/storage-csi/driver-registrar/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/external-attacher/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/external-provisioner/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/csi-controller-rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/controller_service.yaml",
+	)
+	g.cleanup = cleanup
+	if err != nil {
+		framework.Failf("deploying csi hostpath driver: %v", err)
+	}
 }
 
 func (g *gcePDCSIDriver) cleanupCSIDriver() {
 	By("uninstalling gce-pd driver")
-	f := g.f
-	cs := f.ClientSet
-	config := g.config
-	deployGCEPDCSIDriver(cs, config, true /* teardown */, f, g.nodeServiceAccount, g.controllerServiceAccount)
-	csiClusterRoleBindings(cs, config, true /* teardown */, g.controllerServiceAccount, g.controllerClusterRoles)
-	csiClusterRoleBindings(cs, config, true /* teardown */, g.nodeServiceAccount, g.nodeClusterRoles)
-	utils.PrivilegedTestPSPClusterRoleBinding(cs, config.Namespace, true, /* teardown */
-		[]string{g.controllerServiceAccount.Name, g.nodeServiceAccount.Name})
-	role := csiControllerRole(cs, config, true)
-	csiControllerRoleBinding(cs, config, true, role, g.controllerServiceAccount)
-	csiServiceAccount(cs, config, "gce-controller", true /* teardown */)
-	csiServiceAccount(cs, config, "gce-node", true /* teardown */)
+	if g.cleanup != nil {
+		g.cleanup()
+	}
+}
+
+// TODO (?): the storage.csi.image.version and storage.csi.image.registry
+// settings are ignored. We could patch the image definitions.
+// TODO: way too many parameters...
+func patchCSIDeployment(f *framework.Framework, provisioner, driverContainerName, provisionerContainerName, serverNodeName string, object interface{}) {
+	patchVolumes := func(volumes []v1.Volume) {
+		for i := range volumes {
+			volume := &volumes[i]
+			if volume.HostPath != nil {
+				// Update paths like /var/lib/kubelet/plugins/<provisioner>.
+				path := &volume.HostPath.Path
+				if strings.HasSuffix(*path, "/"+provisioner) {
+					*path = *path + "-" + f.UniqueName
+				}
+			}
+		}
+	}
+
+	patchContainers := func(containers []v1.Container) {
+		for i := range containers {
+			container := &containers[i]
+			for e := range container.Args {
+				// Inject test-specific provider name into paths (--kubelet-registration-path).
+				container.Args[e] = strings.Replace(container.Args[e], "/"+provisioner+"/", "/"+provisioner+"-"+f.UniqueName+"/", 1)
+			}
+			// Overwrite driver name resp. provider name
+			// by appending a parameter with the right
+			// value.
+			switch container.Name {
+			case driverContainerName:
+				container.Args = append(container.Args, "--drivername="+provisioner+"-"+f.UniqueName)
+			case provisionerContainerName:
+				container.Args = append(container.Args, "--provisioner="+provisioner+"-"+f.UniqueName)
```

Comment:
Once https://github.com/kubernetes-csi/external-provisioner/pull/142 makes it into 0.4, then this is no longer needed.

Reply:
The new function explicitly allows the container names to be empty, in which case the argument will not be added.